src/peak_event.c

```diff
@@ -0,0 +1,69 @@
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/peak.h"
+#include "jemalloc/internal/peak_event.h"
+
+/*
+ * Update every 100k by default.  We're not exposing this as a configuration
+ * option for now; we don't want to bind ourselves too tightly to any particular
+ * performance requirements for small values, or guarantee that we'll even be
+ * able to provide fine-grained accuracy.
+ */
+#define PEAK_EVENT_WAIT (100 * 1024)
+
+/* Update the peak with current tsd state. */
+void
+peak_event_update(tsd_t *tsd) {
+	uint64_t alloc = tsd_thread_allocated_get(tsd);
+	uint64_t dalloc = tsd_thread_deallocated_get(tsd);
+	peak_t *peak = tsd_peakp_get(tsd);
+	peak_update(peak, alloc, dalloc);
+}
+
+/* Set current state to zero. */
+void
+peak_event_zero(tsd_t *tsd) {
+	uint64_t alloc = tsd_thread_allocated_get(tsd);
+	uint64_t dalloc = tsd_thread_deallocated_get(tsd);
+	peak_t *peak = tsd_peakp_get(tsd);
+	peak_set_zero(peak, alloc, dalloc);
+}
+
+uint64_t
+peak_event_max(tsd_t *tsd) {
+	peak_t *peak = tsd_peakp_get(tsd);
+	return peak_max(peak);
+}
+
+uint64_t
+peak_alloc_new_event_wait(tsd_t *tsd) {
+	return TE_MIN_START_WAIT;
+}
+
+uint64_t
+peak_alloc_postponed_event_wait(tsd_t *tsd) {
+	return TE_MIN_START_WAIT;
+}
+
+uint64_t
```

Comment:
Did you see any warning? It should at least warn because the signature of the event handler function differ from the declarations in `src/thread_event.c`. The event handler function is not supposed to return the wait time.

Reply:
No warnings. Those declarations aren't exposed in the header, so there's no way they could trigger a warning (and of course, it links and runs fine). Will fix.

Reply:
Yeah I see. Do you think I should have structured the declarations in a different way? It's kind of like specifying an interface, but it's now obvious that I wasn't able to make it do its job. Perhaps putting the declarations in `thread_event.h` instead of in `thread_event.c`? There could be some risk of circular dependencies though. For example for the case of your `peak` functionality, there'd be a `tsd` - `thread_event` - `peak` - `tsd` dependency loop, though I think in theory we are able to resolve the loop.